docs/dev_guide/contents/extending_fido.rst

```diff
@@ -0,0 +1,187 @@
+Say, for example, the Wavelength Attr can be passed as an `~astropy.units.Quantity` to the :meth:`~sunpy.net.dataretriever.client.GenericClient.search` but the url may have a different representation for it in its string.
+For such cases, these methods need to be worked out:
+
+* :meth:`~sunpy.net.dataretriever.client.GenericClient.pre_search_hook` which will convert the passed attrs to their representation in the url.
+* :meth:`~sunpy.net.dataretriever.client.GenericClient.post_search_hook` which converts the retrieved metadata from url to the form in which they are desired to be represented in the response table.
+
+It may also be possible that the URL contains the "Attrs" other than time in the directory itself.
+Since scraper doesn't support generating directories that have non-time variables, the :meth:`~sunpy.net.dataretriever.client.GenericClient.search` needs to be overwritten.
+Based on the "Attrs" passed to it, they can be looped to generate the possible patterns for directory and then passed to the scraper.
+:meth:`super().search` can be called per loop.
+
+Examples
+--------
+
+Suppose any file of a data archive can be described by this ``https://some-domain.com/%Y/%m/%d/satname_{SatellitNumber}_{Level}_%y%m%d%H%M%S_{any-2-digit-number}.fits``:
+
+* ``baseurl`` becomes ``r'https://some-domain.com/%Y/%m/%d/satname_(\d){2}_(\d){1}_(\d){12}_(\d){2}\.fits'``.
+
+  Note all variables in the filename are converted to regex that will match any possible value for it.
+  A character enclosed within ``()`` followed by a number enclosed within ``{}`` is used to match the specified number of occurences of that special sequence.
+  For example, ``%y%m%d%H%M%S`` is a six digit variable (2 digits for each) and thus represented by ``r'(\d){12}'``.
+  Note that ``\`` is used to escape the special character ``.``.
+
+* ``pattern`` becomes ``'{}/{year:4d}/{month:2d}{day:2d}/satname_{SatelliteNumber:2d}_{Level:1d}_{:6d}{hour:2d}{minute:2d}{second:2d}_{:2d}.fits'``.
+  Note the sole purpose of ``pattern`` is to extract the information from matched url, using :func:`~sunpy.extern.parse.parse`.
+  So the desired key names for returned dictionary should be written in the pattern within ``{}``, and they should match with the ``attr.__name__``.
+
+*  ``register_values()`` can be written as:
+
+    .. code-block:: python
+
+        @classmethod
+        def register_values(cls):
+
+            from sunpy.net import attrs
+            adict = {
+            attrs.Instrument: [("SatName", "The description of Instrument")],
+            attrs.Physobs: [('some_physobs', 'Phsyobs description')],
+            attrs.Source: [('some_source', 'Source description')],
+            attrs.Provider: [('some_provider', 'Provider description')],
+            attrs.Level: [("1", "Level 1 data"), ("2", "Level 2 data")],
+            attrs.SatelliteNumber: [("16", "Describe it"), ("17", "Describe it")]
+            }
+
+            return adict
```

Comment:
This code block is nearly identical to the one in the example at the end. Do we really need it?